h2o-genmodel/src/main/java/hex/genmodel/algos/gam/GamUtilsISplines.java

```diff
@@ -28,31 +35,149 @@ public static double[] extractKnots(int index, int order, double[] knots) {
+        double[] oneOverDenominator = new double[2];
+        if (order == 1) {
+            oneOverDenominator[0] = 1;
+            oneOverDenominator[1] = 0;
+        } else {
+            double tempDenom = knots[order-1]-knots[0];
+            oneOverDenominator[0] = tempDenom==0
+                    ? 0 : 1.0/tempDenom;
+            tempDenom = knots[order]-knots[1];
+            oneOverDenominator[1] = tempDenom==0
+                    ? 0 : 1.0/tempDenom;
+        }
+        return oneOverDenominator;
+    }
+
     static double[] formNumerator(int order, double[] knots) {
         double[] numerator = new double[2];
         if (order == 1) {
             numerator[0] = 1;
-            numerator[1] = 1;
+            numerator[1] = 0;
         } else {
             numerator[0] = knots[0];
             numerator[1] = knots[order];
         }
         return numerator;
     }
 
-    static double[] formDenominator(int order, double[] knots) {
+    static double[] formDenominatorMSpline(int order, double[] knots) {
         double[] oneOverDenominator = new double[2];
         if (order == 1) {
             oneOverDenominator[0] = 1;
-            oneOverDenominator[1] = 1;
+            oneOverDenominator[1] = 0;
         } else {
-            double tempDenom = knots[order-1]-knots[0];
+            double tempDenom = knots[order]-knots[0];
             oneOverDenominator[0] = tempDenom==0
                     ? 0 : 1.0/tempDenom;
-            tempDenom = knots[order]-knots[1];
+            tempDenom = knots[order]-knots[0];
             oneOverDenominator[1] = tempDenom==0
                     ? 0 : 1.0/tempDenom;
         }
         return oneOverDenominator;
     }
+    
+    /**
+     * This method performs the multiplication of two polynomials where the polynomials are given as a double
+     * array.  This will result in another array which contains the multiplication of the two polynomials.
+     */
+    public static double[] polynomialProduct(double[] coeff1, double[] coeff2) {
+        int firstLen = coeff1.length;
+        int secondLen = coeff2.length;
+        int combinedLen = firstLen*secondLen;
+        int[] firstOrder = arrayInitRange(firstLen, 0);
+        int[] secondOrder = arrayInitRange(secondLen, 0);
+        int highestOrder = firstLen+secondLen-2;
+        double[] combinedCoefficients = new double[highestOrder+1]; // start with order 0
+        List<Double> combinedC = new ArrayList<>();
+        List<Integer> combinedOrder = new ArrayList<>();
+        for (int firstIndex=0; firstIndex < firstLen; firstIndex++) {
+            for (int secondIndex=0; secondIndex < secondLen; secondIndex++) {
+                double tempValue = coeff1[firstIndex]*coeff2[secondIndex];
+                combinedC.add(tempValue);
+                int tempOrder = firstOrder[firstIndex]+secondOrder[secondIndex];
+                combinedOrder.add(tempOrder);
+            }
+        }
```

Comment:
EDIT: Probably not worth it and it might be slower for low order polynomials...

There is asymptotically faster way to do this but I'm not sure if it would be faster in this case - it depends on the size of the coeff array. This is `O(n^2)` but if we look at the polynomial product as a convolution then we can use FFT to do fast convolution to get `O(n log n)`.  Basically it would be something like `IFFT(FFT(coeff1) .* FFT(coeff2))` where `.*` is element-wise product (Hadamard product).

Reply:
Love this!  I was googling FFT toolboxes out there and did not find too many interesting ones.  Since I only do this once to calculate the penalty matrix, it is okay here.  However, in the future, will look into doing this.  Great suggestion!